pkg/client/v1/nimbleos/access_control_record.go

```diff
@@ -0,0 +1,133 @@
+	Lun *int64 `json:"lun,omitempty"`
+	// VolId - Identifier for the volume this access control record applies to.
+	VolId *string `json:"vol_id,omitempty"`
+	// VolName - Name of the volume this access control record applies to.
+	VolName *string `json:"vol_name,omitempty"`
+	// VolAgentType - External management agent type.
+	VolAgentType *NsAgentType `json:"vol_agent_type,omitempty"`
+	// PeId - Identifier for the protocol endpoint this access control record applies to.
+	PeId *string `json:"pe_id,omitempty"`
+	// PeName - Name of the protocol endpoint this access control record applies to.
+	PeName *string `json:"pe_name,omitempty"`
+	// PeLun - LUN (Logical Unit Number) to associate with this protocol endpoint. Valid LUNs are in the 0-2047 range.
+	PeLun *int64 `json:"pe_lun,omitempty"`
+	// SnapId - Identifier for the snapshot this access control record applies to.
+	SnapId *string `json:"snap_id,omitempty"`
+	// SnapName - Name of the snapshot this access control record applies to.
+	SnapName *string `json:"snap_name,omitempty"`
+	// PeIds - List of candidate protocol endpoints that may be used to access the Virtual Volume. One of them will be selected for the access control record. This field is required only when creating an access control record for a Virtual Volume.
+	PeIds []*string `json:"pe_ids,omitempty"`
+	// Snapluns - Information about the snapshot LUNs associated with this access control record. This field is meaningful when the online snapshot can be accessed as a LUN in the group.
+	Snapluns []*NsSnapLunInfo `json:"snapluns,omitempty"`
+	// CreationTime - Time when this access control record was created.
+	CreationTime *int64 `json:"creation_time,omitempty"`
+	// LastModified - Time when this access control record was last modified.
+	LastModified *int64 `json:"last_modified,omitempty"`
+	// AccessProtocol - Access protocol of the volume.
+	AccessProtocol *NsAccessProtocol `json:"access_protocol,omitempty"`
+}
+
+// EncodeAccessControlRecord - Transform AccessControlRecord to accessControlRecord type
+func EncodeAccessControlRecord(request interface{}) (*accessControlRecord, error) {
+	reqAccessControlRecord := request.(*AccessControlRecord)
+	byte, err := json.Marshal(reqAccessControlRecord)
+	objPtr := &accessControlRecord{}
+	err = json.Unmarshal(byte, objPtr)
+	return objPtr, err
+}
+
+// DecodeAccessControlRecord Transform accessControlRecord to AccessControlRecord type
+func DecodeAccessControlRecord(request interface{}) (*AccessControlRecord, error) {
+	reqAccessControlRecord := request.(*accessControlRecord)
+	byte, err := json.Marshal(reqAccessControlRecord)
```

Comment:
Model classes are basically nimble objects. We cannot use non-pointers to attributes because sdk model class has both request/response attributes in the same class and we really dont want to send response attributes to request API call since the nimble array wont accept it. So we decided to use pointers, as nil pointers are dropped by json request. But that created another sdk user visible issue since now user has to create pointers to attributes for all the attributes of the object which is not very intuitive. Thus we introduced encoding/decoding which converts non-pointers to pointers. Hope this helps .

Reply:
I see I'm missing a lot of history.  Thanks for sharing it with me.  I'm still unsure how we suppress the unwanted attributes, though.  Is an "unset" attribute marshalled in such a way that the unmarshalling recognizes it is unset and leaves the pointer nil?  I'd love to see an example.

Do we depend on the SDK user to leave the correct attributes unset?  How do they know which attributes are applicable to which actions?

Reply:
This is a good starting point  https://www.sohamkamani.com/golang/2018-07-19-golang-omitempty/ link. That will be part of documentation. Also if you happen to set the incorrect attribute and nimos will return an error SM_unexpected_arg.

Reply:
Thanks for the pointer.  I understand the idea of using pointers to indicate whether a value is set.  What I'm not following is how we get from the user model to a correct pointer value, nil or non-nil.  Let's suppose we have a user model class called Size that contains an int.  The internal "size" class will contain a *int.  If a "Size" object has a value of "0", will the "size" object contain nil or a pointer to 0?  How does it decide?  Maybe the user really is trying to update an attribute from some other value to 0.

Reply:
I see your point, the value will be nil for size in your case.  Let me see how this can be handled.